pingcap/tiup
# wait for leader to transfer back during rolling restart

Signed-off-by: cosven <yinshaowen241@gmail.com>



### What problem does this PR solve? 

Before, there is a problem:

> Assume there are 3 tikv instances (1,2,3), and the total number of leaders is 300. When tikv1 is restarted, leaders will migrate to 2 and 3, which are 150 respectively. When restarting tikv2, the 150 leaders were migrated to tikv1 and tikv 3 at the same  time. In addition, considering that the newly restarted tikv may not be able to accept the new leader immediately, in this scenario, the leader will be strongly skewed to tikv3.

[This demo](https://github.com/pingcap/tiup/pull/2043) shows this approch can improve performance stability during rolling restart.

Performance degradation
|               | patched | default |
|---------------|---------|---------|
| mid pressure  | 6%      | 17%     |
| high pressure | 8%      | 22%     |

Total rolling restart time
|               | patched  | default  |
|---------------|----------|----------|
| mid pressure  | 135s     | 105~120s |
| high pressure | 165~180s | 150~165s |

![image](https://user-images.githubusercontent.com/4962134/194830647-2df10054-b141-4949-bab2-6c5e3438c0d4.png)


### What is changed and how it works?

This commit let tiup wait for leader to transfer back during rolling restart. In this way, the pressure should be relatively balance between tikv instances during rolling restart.

### Check List 

Tests 

 - [x] Unit test: It seems there is no need.
 - [x] Integration test:  It seems there is no need.
     - cases
       1. case1: tiup cluster do not need to wait when the total region count is small (less than 60).
       2. case2: tiup cluster need to wait until leaders are transfered back when the cluster is relatively big.
     - steps:
       1. Run `tiup cluster reload xxx -R tikv -y` to do rolling restart on one tidb cluster.
       2. Check if the leader count has recovered before next tikv is restarted.

Code changes

Side effects

 - Possible performance regression
   The duration of rolling restart may be longer than before.

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation


Release notes:

```release-note
Reduce the performance degradation during rolling restart.
```
